sync/etcdv3/etcd_test.go

```diff
@@ -277,88 +233,140 @@ func TestWatchWithRevision(t *testing.T) {
+func TestShouldReturnAllValuesWhenWatchingAtCurrentRevision(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 
-	sync := newSync(t)
-	sync.etcdClient.Delete(ctx, "/", etcd.WithPrefix())
+	sync := newSync(t, ctx)
+	sync.cleanup()
 
-	err := sync.Update(ctx, "/path/to/somewhere", "test1")
-	if err != nil {
-		t.Fatalf("unexpected error")
-	}
-	err = sync.Update(ctx, "/path/to/elsewhere", "test2")
-	if err != nil {
-		t.Fatalf("unexpected error")
-	}
-	err = sync.Update(ctx, "/path/notto/elsewhere", "test3")
-	if err != nil {
-		t.Fatalf("unexpected error")
-	}
+	path := "/path/to/watch/with/revision"
 
-	nodes, err := sync.Fetch(ctx, "/path")
-	if err != nil {
-		t.Fatalf("unexpected error %s", err.Error())
-	}
-	if nodes.Key != "/path" {
-		t.Fatalf("expected node to has key /path has %s instead", nodes.Key)
-	}
-	if len(nodes.Children) != 2 {
-		t.Fatalf("expected 2 node has %d", len(nodes.Children))
-	}
-	if nodes.Children[0].Key != "/path/notto" {
-		t.Fatalf("expected node to has key /path/notto has %s instead", nodes.Children[0].Key)
+	firstRevision := sync.mustPut(path+"/first", `{"version": "1"}`)
+	secondRevision := sync.mustPut(path+"/second", `{"version": "2"}`)
+
+	responseChan := sync.Watch(ctx, path, gohan_sync.RevisionCurrent)
+
+	resp := <-responseChan
+	if resp.Key != path+"/first" || resp.Data["version"].(string) != "1" || resp.Revision != firstRevision {
+		t.Fatalf("mismatch response: %+v, expecting version==1, revision==%d", resp, firstRevision)
 	}
-	if len(nodes.Children[0].Children) != 1 {
-		t.Fatalf("expected 1 node has %d", len(nodes.Children[0].Children))
+
+	resp = <-responseChan
+	if resp.Key != path+"/second" || resp.Data["version"].(string) != "2" || resp.Revision != secondRevision {
+		t.Fatalf("mismatch response: %+v, expecting version==2, revision==%d", resp, secondRevision)
 	}
-	node0 := nodes.Children[0].Children[0]
-	if node0.Value != "test3" || node0.Key != "/path/notto/elsewhere" || len(node0.Children) != 0 {
-		t.Fatalf("incorrect value for node %+v", node0)
+}
+
+func TestShouldStartWatchingAtSpecifiedRevision(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	sync := newSync(t, ctx)
+	sync.cleanup()
+
+	path := "/path/to/watch/with/starting/revision"
+
+	firstRevision := sync.mustPut(path, `{"version": "1"}`)
+	secondRevision := sync.mustPut(path, `{"version": "2"}`)
+
+	responseChan := sync.Watch(ctx, path, firstRevision-1)
+
+	resp := <-responseChan
+	if resp.Key != path || resp.Data["version"].(string) != "1" || resp.Revision != firstRevision {
+		t.Fatalf("mismatch response: %+v, expecting version==1, revision==%d", resp, firstRevision)
 	}
-	if nodes.Children[1].Key != "/path/to" {
-		t.Fatalf("expected node to has key /path/to has %s instead", nodes.Children[1].Key)
+
+	resp = <-responseChan
+	if resp.Key != path || resp.Data["version"].(string) != "2" || resp.Revision != secondRevision {
```

Comment:
We seem to always check for key, version and revision. Can we extract this pattern into a function? `verifyResponse(resp, path, "2", secondRevision)`?

Reply:
sure, changed